server/libs/storage/src/storage.service.ts

```diff
@@ -150,4 +150,16 @@ export class StorageService {
 
     return template(substitutions);
   }
+
+  public async shouldMigrate(asset: AssetEntity, filename: string): Promise<boolean> {
+    const source = asset.originalPath;
+    const ext = path.extname(source).split('.').pop() as string;
+    const sanitized = sanitize(path.basename(filename, `.${ext}`));
+    const rootPath = path.join(APP_UPLOAD_LOCATION, asset.userId);
+    const storagePath = this.render(this.storageTemplate, asset, sanitized, ext);
+    const fullPath = path.normalize(path.join(rootPath, storagePath));
+    const shouldBeDestination = `${fullPath}.${ext}`;
+    const isFileExist = await this.checkFileExist(shouldBeDestination);
+    return !isFileExist;
```

Comment:
Instead of comparing "dest" to file on disk, can you just compare it against "source"? Specifically if source starts with full path (everything in the template, but not the duplicate stuff or the extension) then you know it already follows the path of the template. Basically if the source is the same as the destination (minus the optional  duplicate portion) there is no need to migrate it.

Reply:
I was going to compare the string, but since there is a helper method, I used it for readability.

Reply:
But those are two different things. The file existing doesn't necessarily mean it's for this asset.

Reply:
I remember that overtime, there was an implementation that write the original path as `./upload/...` instead of `upload/...` so comparing string would be ugly here